kore/src/Kore/Step/Simplification/Condition.hs

```diff
@@ -120,49 +123,71 @@ simplify SubstitutionSimplifier{simplifySubstitution} sideCondition =
-            [Predicate RewritingVariableName] ->
-            StateT
-                (Condition RewritingVariableName)
-                (LogicT simplifier)
-                ()
-        worker [] = return ()
-        worker (pred' : rest) = do
-            condition <- State.get
-            let otherConds =
-                    SideCondition.andCondition
-                        sideCondition
-                        (mkOtherConditions condition rest)
-            result <-
-                simplifyPredicate otherConds pred'
-                    & lift
-            State.put (Condition.andCondition condition result)
-            worker rest
-
-        mkOtherConditions (Condition.toPredicate -> alreadySimplified) rest =
-            from @_ @(Condition _)
-                . MultiAnd.toPredicate
-                . MultiAnd.make
-                $ alreadySimplified : rest
+simplifyPredicates sideCondition original = do
+    let predicates =
+            SideCondition.simplifyConjunctionByAssumption original
+                & fst . extract
+    simplified <-
+        simplifyPredicatesWithAssumptions
+            sideCondition
+            (toList predicates)
+    let simplifiedPredicates =
+            from @(Condition _) @(MultiAnd (Predicate _)) simplified
+    if original == simplifiedPredicates
+        then return (Condition.markSimplified simplified)
+        else simplifyPredicates sideCondition simplifiedPredicates
 
-        markConjunctionSimplified =
-            return . Lens.over (field @"predicate") Predicate.markSimplified
+{- | Simplify a conjunction of predicates by simplifying each one
+under the assumption that the others are true.
+-}
+simplifyPredicatesWithAssumptions ::
+    forall simplifier.
+    MonadSimplify simplifier =>
+    SideCondition RewritingVariableName ->
+    [Predicate RewritingVariableName] ->
+    LogicT
+        simplifier
+        (Condition RewritingVariableName)
+simplifyPredicatesWithAssumptions _ [] = return Condition.top
+simplifyPredicatesWithAssumptions sideCondition predicates@(_ : rest) = do
+    let predicatesWithUnsimplified =
+            zip predicates $
+                scanr ((<>) . MultiAnd.singleton) MultiAnd.top rest
+    State.execStateT
+        ( traverse
+            simplifyWithAssumptions
+            predicatesWithUnsimplified
+        )
+        Condition.top
+  where
+    simplifyWithAssumptions ::
+        ( Predicate RewritingVariableName
+        , MultiAnd (Predicate RewritingVariableName)
+        ) ->
+        StateT
+            (Condition RewritingVariableName)
+            (LogicT simplifier)
+            ()
+    simplifyWithAssumptions (predicate, unsimplifiedSideCond) = do
+        simplifiedSideCond <- State.get
+        let otherSideConds =
+                SideCondition.addPredicates
+                    ( from @_ @(MultiAnd (Predicate _)) simplifiedSideCond
+                        <> unsimplifiedSideCond
+                    )
+                    sideCondition
```

Comment:
Why not `SideCondition.addPredicates unsimplifiedSideCond (simplifiedSideCond <> sideCondition)`?